plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -393,6 +463,46 @@ public ConnectorPageSource createPageSource(
                 getClass().getClassLoader());
     }
 
+    private ConnectorPageSource createIcebergSystemPageSource(
+            List<ColumnHandle> columns,
+            IcebergTableHandle table,
+            ConnectorSession session,
+            ConnectorTransactionHandle transaction,
+            Function<Table, IcebergSystemTable> systemTableCreator)
+    {
+        Table icebergTable = catalogFactory.create(session.getIdentity()).loadTable(session, table.getSchemaTableName());
+        IcebergSystemTable systemTable = systemTableCreator.apply(icebergTable);
+        List<ColumnMetadata> tableColumns = systemTable.getTableMetadata().getColumns();
+
+        Map<String, Integer> columnsByName = new HashMap<>();
+        for (int i = 0; i < tableColumns.size(); i++) {
+            ColumnMetadata column = tableColumns.get(i);
+            if (columnsByName.put(column.getName(), i) != null) {
+                throw new TrinoException(GENERIC_INTERNAL_ERROR, "Duplicate column name: " + column.getName());
+            }
+        }
+
+        ImmutableList.Builder<Integer> userToSystemFieldIndex = ImmutableList.builder();
+        for (ColumnHandle column : columns) {
+            String columnName = ((IcebergColumnHandle) column).getName();
+
+            Integer index = columnsByName.get(columnName);
+            if (index == null) {
+                throw new TrinoException(GENERIC_INTERNAL_ERROR, format("Column does not exist: %s.%s", table.getSchemaTableName(), columnName));
+            }
+
+            userToSystemFieldIndex.add(index);
+        }
+
+        // TODO verify whether IcebergColumnHandle should transport for the system tables the constraint (same as SystemTableHandle) and replace TupleDomain.all() with the constraint
```

Comment:
I would benefit of some guidance here.

cc @findepi @electrum

Reply:
(discussed offline)